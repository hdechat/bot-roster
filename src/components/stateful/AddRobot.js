import React, { Component } from 'react';

const initialState = {
  category: '',
  firstName: '',
  lastName: '',
  speed: 0,
  strength: 0,
  agility: 0,
};

class AddRobot extends Component {
  constructor(props) {
    super(props);

    this.state = initialState;
  };

  updateUserInputText = (event) => {
    this.setState({ [event.target.name]: event.target.value });
  };

  updateUserInputNumber = (event) => {
    this.setState({ [event.target.name]: parseInt(event.target.value) })
  };

  addToTeam = (event) => {
    event.preventDefault();
    this.props.addRobotToTeam(this.state);
    this.clearStateValues();
  };

  clearStateValues = () => {
    this.setState({ ...this.state, ...initialState });
  }

  render() {
    return (
      <form onSubmit={this.addToTeam}>
        <select required name="category" onChange={this.updateUserInputText}>
          <option value="">Select a Category</option>
          <option value="starters">starters</option>
          <option value="subs">subs</option>
        </select>
        <input
          required
          value={this.state.firstName}
          name="firstName"
          onChange={this.updateUserInputText} 
          placeholder="first name"
          minLength="2" />
        <input
          required
          value={this.state.lastName} 
          name="lastName" 
          onChange={this.updateUserInputText} 
          placeholder="last name" 
          minLength="2"/>
        <input 
          type="number"
          value={this.state.speed} 
          name="speed"
          onChange={this.updateUserInputNumber} 
          placeholder="speed" />
        <input 
          type="number"
          value={this.state.strength} 
          name="strength" 
          onChange={this.updateUserInputNumber} 
          placeholder="strength" />
        <input 
          type="number"
          value={this.state.agility} 
          name="agility" 
          onChange={this.updateUserInputNumber} 
          placeholder="agility" />
        <button type="submit">Add Robot To Team</button>
      </form>
    );
  };
};

export default AddRobot;
