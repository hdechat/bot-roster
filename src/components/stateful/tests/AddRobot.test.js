import React from 'react';
import AddRobot from '../AddRobot';
import { shallow, mount } from 'enzyme';

describe('AddRobot', () => {
  let wrapper;
  let mockAddRobotToTeam;
  let mockState = {
    category: '',
    firstName: '',
    lastName: '',
    speed: 0,
    strength: 0,
    agility: 0  
  }

  beforeEach(() => {
    mockAddRobotToTeam = jest.fn();

    wrapper = shallow(<AddRobot addRobotToTeam={mockAddRobotToTeam} />)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('UPDATE_USER_INPUT_TEXT', () => {
    it('should update state', () => {
      const mockEvent = {target: {
        name: 'category',
        value: 'subs'
      }}

      const expectedState = {
        category: 'subs',
        firstName: '',
        lastName: '',
        speed: 0,
        strength: 0,
        agility: 0  
      };   

      wrapper.instance().updateUserInputText(mockEvent);

      expect(wrapper.state()).toEqual(expectedState);
    });
  });

describe('UPDATE_USER_INPUT_NUMBER', () => {
    it('should update state with number', () => {
      const mockEvent = {target: {
        name: 'speed',
        value: '5'
      }}

      const expectedState = {
        category: '',
        firstName: '',
        lastName: '',
        speed: 5,
        strength: 0,
        agility: 0  
      };   

      wrapper.instance().updateUserInputNumber(mockEvent);

      expect(wrapper.state()).toEqual(expectedState);
    });
  });

  describe('ADD_TO_TEAM', () => {
    const mockEvent = {preventDefault: jest.fn()}
    
    it('calls addRobotToTeam with correct params', () => {
      wrapper.instance().addToTeam(mockEvent);

      expect(mockAddRobotToTeam).toHaveBeenCalledWith(mockState);
    });

    it('calls clearStateValues', () => {
      wrapper = mount(<AddRobot addRobotToTeam={mockAddRobotToTeam} />)
      const spy = spyOn(wrapper.instance(), 'clearStateValues');

      wrapper.instance().forceUpdate();
      wrapper.instance().addToTeam(mockEvent);

      expect(spy).toHaveBeenCalled();
    }); 
  });

  describe('CLEAR_STATE_VALUES', () => {
    it('clears state values', () => {
      const mockUpdatedState = {
        category: 'subs',
        firstName: 'Jhonen',
        lastName: 'Vasquez',
        speed: 10,
        strength: 40,
        agility: 50  
      };

      wrapper.setState(mockUpdatedState)
      wrapper.instance().clearStateValues();

      expect(wrapper.state()).toEqual(mockState);
    });
  });

  it('renders the expected elements', () => {
    expect(wrapper.find('form').length).toEqual(1);
    expect(wrapper.find('select').length).toEqual(1);
    expect(wrapper.find('option').length).toEqual(3);
    expect(wrapper.find('input').length).toEqual(5);
    expect(wrapper.find('button').length).toEqual(1);
  });

  it('calls addToTeam on form submit', () => {
    wrapper = mount(<AddRobot addRobotToTeam={mockAddRobotToTeam} />)
    const spy = spyOn(wrapper.instance(), 'addToTeam');

    wrapper.instance().forceUpdate();
    wrapper.find('form').simulate('submit');

    expect(spy).toHaveBeenCalled();
  });
});
  
