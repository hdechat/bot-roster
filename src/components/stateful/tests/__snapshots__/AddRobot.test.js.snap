// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRobot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddRobot
    addRobotToTeam={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <select
          name="category"
          onChange={[Function]}
          required={true}
>
          <option
                    value=""
          >
                    Select a Category
          </option>
          <option
                    value="starters"
          >
                    starters
          </option>
          <option
                    value="subs"
          >
                    subs
          </option>
</select>,
        <input
          minLength="2"
          name="firstName"
          onChange={[Function]}
          placeholder="first name"
          required={true}
          value=""
/>,
        <input
          minLength="2"
          name="lastName"
          onChange={[Function]}
          placeholder="last name"
          required={true}
          value=""
/>,
        <input
          name="speed"
          onChange={[Function]}
          placeholder="speed"
          type="number"
          value={0}
/>,
        <input
          name="strength"
          onChange={[Function]}
          placeholder="strength"
          type="number"
          value={0}
/>,
        <input
          name="agility"
          onChange={[Function]}
          placeholder="agility"
          type="number"
          value={0}
/>,
        <button
          type="submit"
>
          Add Robot To Team
</button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
>
              Select a Category
</option>,
            <option
              value="starters"
>
              starters
</option>,
            <option
              value="subs"
>
              subs
</option>,
          ],
          "name": "category",
          "onChange": [Function],
          "required": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a Category",
              "value": "",
            },
            "ref": null,
            "rendered": "Select a Category",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "starters",
              "value": "starters",
            },
            "ref": null,
            "rendered": "starters",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "subs",
              "value": "subs",
            },
            "ref": null,
            "rendered": "subs",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "minLength": "2",
          "name": "firstName",
          "onChange": [Function],
          "placeholder": "first name",
          "required": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "minLength": "2",
          "name": "lastName",
          "onChange": [Function],
          "placeholder": "last name",
          "required": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "speed",
          "onChange": [Function],
          "placeholder": "speed",
          "type": "number",
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "strength",
          "onChange": [Function],
          "placeholder": "strength",
          "type": "number",
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "agility",
          "onChange": [Function],
          "placeholder": "agility",
          "type": "number",
          "value": 0,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add Robot To Team",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Add Robot To Team",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <select
            name="category"
            onChange={[Function]}
            required={true}
>
            <option
                        value=""
            >
                        Select a Category
            </option>
            <option
                        value="starters"
            >
                        starters
            </option>
            <option
                        value="subs"
            >
                        subs
            </option>
</select>,
          <input
            minLength="2"
            name="firstName"
            onChange={[Function]}
            placeholder="first name"
            required={true}
            value=""
/>,
          <input
            minLength="2"
            name="lastName"
            onChange={[Function]}
            placeholder="last name"
            required={true}
            value=""
/>,
          <input
            name="speed"
            onChange={[Function]}
            placeholder="speed"
            type="number"
            value={0}
/>,
          <input
            name="strength"
            onChange={[Function]}
            placeholder="strength"
            type="number"
            value={0}
/>,
          <input
            name="agility"
            onChange={[Function]}
            placeholder="agility"
            type="number"
            value={0}
/>,
          <button
            type="submit"
>
            Add Robot To Team
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
>
                Select a Category
</option>,
              <option
                value="starters"
>
                starters
</option>,
              <option
                value="subs"
>
                subs
</option>,
            ],
            "name": "category",
            "onChange": [Function],
            "required": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a Category",
                "value": "",
              },
              "ref": null,
              "rendered": "Select a Category",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "starters",
                "value": "starters",
              },
              "ref": null,
              "rendered": "starters",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "subs",
                "value": "subs",
              },
              "ref": null,
              "rendered": "subs",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "minLength": "2",
            "name": "firstName",
            "onChange": [Function],
            "placeholder": "first name",
            "required": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "minLength": "2",
            "name": "lastName",
            "onChange": [Function],
            "placeholder": "last name",
            "required": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "speed",
            "onChange": [Function],
            "placeholder": "speed",
            "type": "number",
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "strength",
            "onChange": [Function],
            "placeholder": "strength",
            "type": "number",
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "agility",
            "onChange": [Function],
            "placeholder": "agility",
            "type": "number",
            "value": 0,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Robot To Team",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Add Robot To Team",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
